### PR TITLE
[FIX] remove the possibility to create a circular dependency

### PR DESCRIPTION
**Pull request by loco-odoo**
[FIX] website: prevent circular dependency of duplicated fields

Steps to reproduce the bug:
- Drop a "Form" snippet on the website.
- Add three fields and rename them by "a", "b" and "b".
- Change the conditional visibility of the first "b" and make it depend
on "a".
- Change the conditional visibility of "a" and make it depend on "b".
- Save.

-> Traceback "Maximum call stack exceeded" appears.

A field with a conditional visibility is visible if at at least one
field with the dependency name is visible. The problem is that in our
case, it exists a circular dependency between "a" and one of the "b"
leading to an infinite loop during this check. Before this commit all
the fields of the form were checked and all the labels of the fields
that do not create a circular dependency were proposed in the file
visibility selector. The problem is that in our case, one of the "b"
field does not create a circular dependency while the other does. To
solve the problem, the `_recursiveFindCircular()` function has been
adapted in order to not propose a label that would create a circular
dependency in the file visibility selector.

task-3291044

-----------------------------------------------------------------------------------------------------------------------------------------------------------

[FIX] website: prevent circular dependency of renamed fields

Steps to reproduce the bug:
- Drop a "Form" snippet on the website.
- Change the "Label" of the first field (and put it to "test" for
example).
- Select the last field and change its visibility so that it depends on
the first field.
- Change the "Label" of the selected field and put it to the same than
the first one ("test").
- Save.

-> Traceback "Maximum call stack exceeded" appears.

At the end of the procedure, the last field visibility depends on itself
and a circular visibility dependency is created. To solve the problem,
this commit checks that the renamed field does not bring a circular
dependency while updating the dependencies. If it is the case, the
problematic dependency is deleted.
To resolve this bug, the `visitedFields` set has been introduced. Its
goal is to register the already visited fields to not enter an infinite
check loop. Let's take an example to illustrate this: Imagine there is a
form of type A->B->C->D. In this form, the field "A" depends on "B" that
depends on "C" that depends on "D". Imagine you rename "D" by "B". You
now have A->B->C->B. The system will check that the renamed field does
not create a circular dependency. To do so, it will apply
`_recursiveFindCircular` with "A" as the `targetFieldEl` and "B" as the
`dependentFieldEl`. Because there is a circular dependency between "C"
and "B", the system would enter in an infinite loop. Note that we do not
notify the system if such indirect circular dependency has been found as
it will be detected as a direct circular dependency  while the system
will apply `_recursiveFindCircular` with `C` as the `targetFieldEl`.

task-3291044

-----------------------------------------------------------------------------------------------------------------------------------------------------------

[FIX] website: re-render the visibility selector on field label change

Steps to reproduce the bug:
- Add a form on the website.
- Add two fields on the form ("A" and "B").
- Change the conditional visibility of "B" and put it to "Visible only
if".
- Change the label of "B" and set it to "A".
- Try to change the dependent field.

-> "A" is still on the list of the eligible dependent fields. It should
not be the case as setting this field as a dependent field would create
a circular dependency.

To solve the problem, the file visibility selector is rerendered after
modifying a field label. Because of the `_findCircular()` check at the
rerendering, the problematic field will not be displayed in the list of
of the eligible dependent fields.

task-3291044

### CODE REVIEW

File: addons/website/static/src/snippets/s_website_form/options.js [hunk 1]
```diff
@@ -1239,14 +1240,24 @@ options.registry.WebsiteFieldEditor = FieldEditor.extend({
...
-                return false;
+            const dependentFieldName = this._getFieldName(el);
+            // Get all the fields that have the same label than the input.
+            let fieldEls = Array.from(this.formEl
+                .querySelectorAll(`.s_website_form_input[name="${dependentFieldName}"]`))
+                .map((el) => el.closest(".s_website_form_field"));
+            // Filter the array to remove the duplicate fields. This could
+            // happen if the field has multiple inputs ("Multiple Checkboxes" for
+            // example.)
+            fieldEls = fieldEls.filter((fieldEl, index) => fieldEls.indexOf(fieldEl) === index);
```
Comment (bvr-odoo) [inline addons/website/static/src/snippets/s_website_form/options.js]:
```suggestion
fieldEls = [...new Set(fieldEls)];
```

File: addons/website/static/src/snippets/s_website_form/options.js [hunk 2]
```diff
@@ -1239,14 +1240,24 @@ options.registry.WebsiteFieldEditor = FieldEditor.extend({
...
+            const fieldName = this._getFieldName();
+            for (const fieldEl of fieldEls) {
+                const dependencyInputEl = this._getDependencyEl(fieldEl);
```
Comment (bvr-odoo) [inline addons/website/static/src/snippets/s_website_form/options.js]: You don't need to do `_getDependencyEl` if `fieldEl.dataset.visibilityDependency === fieldName` so you could avoid to do it.

**Comment by qsm-odoo** [comment]
@robodoo delegate=detrouxdev 
@detrouxdev

File: addons/website/static/src/snippets/s_website_form/options.js [hunk 3]
```diff
@@ -1239,14 +1240,26 @@ options.registry.WebsiteFieldEditor = FieldEditor.extend({
...
+            let fieldEls = Array.from(this.formEl
+                .querySelectorAll(`.s_website_form_input[name="${dependentFieldName}"]`))
+                .map((el) => el.closest(".s_website_form_field"));
+            // Filter the array to remove the duplicate fields. This could
+            // happen if the field has multiple inputs ("Multiple Checkboxes" for
+            // example.)
+            fieldEls = [...new Set(fieldEls)];
+            const fieldName = this._getFieldName();
+            for (const fieldEl of fieldEls) {
+                if (fieldEl.dataset.visibilityDependency === fieldName) {
```
Comment (detrouxdev) [inline addons/website/static/src/snippets/s_website_form/options.js]: [Set can be iterated on](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/@@iterator), so there's no need to convert it to an array, you can just
```js
fieldEls = new Set(fieldEls);
for (const fieldEl of fieldEls) {
    ...
}
```

File: addons/website/static/src/snippets/s_website_form/options.js [hunk 4]
```diff
@@ -985,6 +985,13 @@ options.registry.WebsiteFieldEditor = FieldEditor.extend({
      * Set the name of the field on the label
      */
     setLabelText: function (previewMode, value, params) {
+        if (!previewMode && this._getDependencyEl() &&
+            this._recursiveFindCircular(this._getDependencyEl().closest(".s_website_form_field"))) {
+            // A circular dependency can appear if for example a field (A)
+            // depends on another (B) and the user rename "A" by "B". In this
+            // case, remove the problematic circular dependency.
+            this._deleteConditionalVisibility(this.$target[0]);
+        }
```
Comment (detrouxdev) [inline addons/website/static/src/snippets/s_website_form/options.js:994]: Just a suggestion, though I think storing dependencyEl is probably a nice thing :D
```js
const dependencyEl = this._getDependencyEl():
if (
    !previewMode 
    && dependencyEl 
    && this._recursiveFindCircular(dependencyEl.closest(".s_website_form_field"))
 ) {
     ...
 }
 ```

**Review by detrouxdev — CHANGES_REQUESTED** [review]

File: addons/website/static/src/snippets/s_website_form/options.js (hunk 4 above)
Comment (loco-odoo) [reply addons/website/static/src/snippets/s_website_form/options.js]: @detrouxdev , Indeed, thanks for reporting! I adapted the PR accordingly :+1:

**Comment by detrouxdev** [comment]
@loco-odoo one last change: 😅 could you rename your commits please? :pray: I feel like the numbers don't make it clear what you're actually fixing. Something like:
- [FIX] website: remove the possibility to create a circular dependency 1 => [FIX] website: prevent circular dependency of duplicate fields
- [FIX] website: remove the possibility to create a circular dependency 2 => [FIX] website: prevent circular dependency of renamed fields

And you can probably squash commit 2 and commit 3 as they both happen when they are renamed. Or add another title like `[FIX] website: avoid circular dependency by re-rendering form options`

You can also squash all three and explain all the steps in one big commit message, I'll let you decided that. Once all of that is done, those fixes will be ready to be shipped 🚀

**Comment by loco-odoo** [comment]
> could you rename your commits please? 

@detrouxdev , indeed, there were not precise enough, I have just renamed them :+1:

**Comment by detrouxdev** [comment]
@loco-odoo thanks for the fixes and for applying my suggestions 🙏🏻 

@robodoo r+ rebase-ff

**Pull request closed without merging**